src/ppx_deriving_yojson.ml

```diff
@@ -316,17 +312,18 @@ let ser_type_of_decl ~options ~path:_ type_decl =
                        (fun var -> [%type: [%t var] -> Yojson.Safe.t]) type_decl in
   polymorphize [%type: [%t typ] -> Yojson.Safe.t]
 
-let ser_str_of_record ~loc varname labels =
+let ser_str_of_record quoter ~loc varname labels =
   let fields =
     labels |> List.mapi (fun _i { pld_loc = loc; pld_name = { txt = name }; pld_type; pld_attributes } ->
       let field  = Exp.field (evar varname) (mknoloc (Lident name)) in
       let result = [%expr [%e str (attr_key name pld_attributes)],
-                    [%e ser_expr_of_typ @@ type_add_attrs pld_type pld_attributes] [%e field]] in
+                    [%e ser_expr_of_typ quoter @@ type_add_attrs pld_type pld_attributes] [%e field]] in
       match attr_default (pld_type.ptyp_attributes @ pld_attributes) with
       | None ->
           [%expr [%e result] :: fields]
       | Some default ->
-          [%expr if [%e field] = [%e default] then fields else [%e result] :: fields])
+          let default = [%expr ([%e default] : [%t pld_type])] in
+          [%expr if [%e field] = [%e Ppx_deriving.quote ~quoter default] then fields else [%e result] :: fields])
```

Comment:
Here you are using the quoter, but also introducing a type annotation. Can you comment on why you added the type annotation (is it necessary in some cases?).

(If an annotation is required here, why is it not also required in the pattern-matching in `var_s_exn_fun` below?)

Reply:
This is related to my previous comment: https://github.com/ocaml-ppx/ppx_deriving_yojson/pull/140#issuecomment-962625900. It allows the following (added in tests):
https://github.com/ocaml-ppx/ppx_deriving_yojson/blob/e4bd570e18ee99de6acdcba5bc18a48adfda3990/src_test/test_ppx_yojson.ml#L555
Without the type annotation one would have to write `[@default Result.MyResult]` instead, although before quoting this wasn't necessary. Quoting makes the type annotation necessary because otherwise there's an intermediate closure that gets applied, which I think doesn't allow such type-based lookup of the constructor from a different scope.

Reply:
Thanks for the clarification, that makes sense. After thinking more about it, I believe that similar annotation would not be necessary on the use of `Result.Ok` etc. in the `var_s_exn_fun` definition below, and I also checked that all uses of `attr_default` are annotated as expected.